src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,784 @@
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, 0L, length);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.tail(length);
+                    } else {
+                        long startAt = decryption.calculateTail();
+                        getOptions.startAt(startAt);
+                    }
+                } else if (ranges[1].isEmpty()) {
+                    // clear current ranges to avoid multiple ranges
+                    getOptions.getRanges().clear();
+
+                    // handle to read from an offset till the end
+                    long offset = Long.parseLong(ranges[0]);
+                    decryption =
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, offset, -1L);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.startAt(offset);
+                    } else {
+                        long startAt = decryption.getStartAt();
+                        getOptions.startAt(startAt);
+                    }
+                } else {
+                    // clear current ranges to avoid multiple ranges
+                    getOptions.getRanges().clear();
+
+                    // handle to read from an offset a specific amount of bytes
+                    long offset = Long.parseLong(ranges[0]);
+                    long end = Long.parseLong(ranges[1]);
+                    long length = end - offset + 1;
+                    decryption =
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, offset, length);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.range(offset, end);
+                    } else {
+                        long startAt = decryption.getStartAt();
+                        long endAt = decryption.calculateEndAt(end);
+                        getOptions.range(startAt, endAt);
+                    }
+                }
+            }
+
+            // initialize the decryption if no ranges were requested
+            if (decryption == null) {
+                decryption =
+                    new Decryption(secretKey, delegate(), containerName,
+                        blobName, 0L, -1L);
+            }
+
+            return decryptBlob(decryption, containerName,
+                delegate().getBlob(containerName, blobName, getOptions));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public String putBlob(String containerName, Blob blob) {
+        return delegate().putBlob(containerName,
```

Comment:
It would be nice to tee the input stream through the MD5 hash, so that the returned ETag would match the content rather than the encrypted content. Otherwise common clients like s3cmd will break, e.g.:
```
DEBUG: signature-v4 headers: {'content-length': '6943', 'content-type': 'text/plain', 'x-amz-meta-s3cmd-attrs': 'atime:1642891263/ctime:1635210219/gid:1000/gname:timur/md5:d851162cc8f800d496c800efee338e89/mode:33188/mtime:1635210219/uid:1000/uname:timur', 'x-amz-storage-class': 'STANDARD', 'x-amz-date': '20220122T230420Z', 'Authorization': 'AWS4-HMAC-SHA256 Credential=test/20220122/us-east-1/s3/aws4_request,SignedHeaders=content-length;content-type;host;x-amz-content-sha256;x-amz-date;x-amz-meta-s3cmd-attrs;x-amz-storage-class,Signature=84b53372268cbdd61ab664d1f0e6c0b73aff05950d1cc517efba241b1082ef6c', 'x-amz-content-sha256': '08ad980cc2bf1b9ff3ff329ca6402eeb2a0131d17b590169e987b8dbd1060366'}
DEBUG: get_hostname(timur-s3): localhost:10080
DEBUG: ConnMan.get(): re-using connection: http://localhost:10080#1
DEBUG: format_uri(): /timur-s3/README.md
 6943 of 6943   100% in    0s     3.27 MB/sDEBUG: ConnMan.put(): connection put back to pool (http://localhost:10080#2)
DEBUG: Response:
{'data': b'',
 'headers': {'content-length': '0',
             'date': 'Sat, 22 Jan 2022 23:04:20 GMT',
             'etag': '"5405364c01976e87a31e476624d0dfeb"',
             'server': 'Jetty(9.4.41.v20210516)',
             'x-amz-request-id': '4442587FB7D0A2F9'},
 'reason': 'OK',
 'size': 6943,
 'status': 200}
 6943 of 6943   100% in    0s    25.79 KB/s  done
DEBUG: MD5 sums: computed=d851162cc8f800d496c800efee338e89, received=5405364c01976e87a31e476624d0dfeb
WARNING: MD5 Sums don't match!
```